core/trino-spi/src/main/java/io/trino/spi/block/VariableWidthBlockEncoding.java

```diff
@@ -68,7 +68,8 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
         boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
 
         int blockSize = sliceInput.readInt();
-        Slice slice = sliceInput.readSlice(blockSize);
+        Slice slice = Slices.allocate(blockSize);
```

Comment:
This implies data copying. In memory connector we just use `page.compact();` (see https://github.com/trinodb/trino/blob/f13283bcc5a7507fa5396cdec9afbda93d642ee6/plugin/trino-memory/src/main/java/io/trino/plugin/memory/MemoryPagesStore.java#L69)

Reply:
Are you suggesting me to compact page before return here?[https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeUtil.java#L61](url)

Reply:
It's the only place when in block decoding a view is preferred to a copy. It feels like it is more natural to create a copy here. @sopel39 Do you think there's a practical reason why it is better to compact vs creating a copy explicitly here?

Reply:
> Are you suggesting me to compact page before return here?

I'm suggesting to `compact` if you mean to store page. This is exactly what we do in `OrderByOperator`, `PartitioningExchanger`, `PagesIndex` and others.

> @sopel39 Do you think there's a practical reason why it is better to compact vs creating a copy explicitly here?

You don't need that extra copy if the data is going to be consumed immediately (e.g. aggregation)

Reply:
It looks like the uncompacted pages are somehow being retained for longer than expected. @linzebing what is the code path that retains those pages?

Reply:
Looks that we need to compact here https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchangeSinkOperator.java#L147-L154

Reply:
@arhimondr @sopel39 : there are at least five operators retaining the reference without compacting pages:
```
WorkProcessorOperatorAdapter
TableWriterOperator
TaskOutputOperator
LocalExchangeSinkOperator
DistinctLimitOperator
```

Another issue is that, is it expected that `compact` doesn't compact the `dictionary` field of a `DictionaryBlock`?
<img width="1066" alt="image" src="https://user-images.githubusercontent.com/9404831/156468221-d76d5755-d1bd-4519-9f89-df382e9cd60c.png">

Reply:
I'm a little bit worried that it might keep stumbling on bugs related to this problem if we try to address it by applying compaction on the operator level

Reply:
The reason we don't compact or eagerly copy it that is can be really expensive.  This is the same reason we don't compact dictionaries.  Also there are methods that will attempt to dedupe things, but we don't use those because they had a big performance impact.

All of that said, I'm not saying "don't do this".  I'm saying "be careful" and check the performance.

Reply:
>  Also there are methods that will attempt to dedupe things, but we don't use those because they had a big performance impact.

By that @dain means using something like `io.trino.operator.project.PageProcessor.ProjectSelectedPositions#updateRetainedSize` on `Page` level.

>  I think this is because compact doesn't compact the dictionary field of a DictionaryBlock, as I mentioned above.

I think we should compact dictionaries on `io.trino.spi.block.DictionaryBlock#compact`

Reply:
I'll jump in with yet another potential solution here:
You could re-frame this as "`VariableWidthBlock` is only block encoding that doesn't copy memory into a new structure _again_". It would be a bigger change, but if the deserialization operated directly on an `InputStreamSliceInput` in the  first place instead of eagerly copying the whole serialized page into a `Slice`- then this wouldn't be adding unnecessary overheads.

I've looked into trying to do something like that before, and the problems I saw were solvable but included:
- Finding a clean way to still be able to compute the page checksum on the input stream (and having a way to get the current digest and reset for the next logical page at some middle point in the buffer)
- Verifying that moving the deserialization point into the I/O layers (eg: exchange client, unspilling thread) wouldn't degrade performance in unexpected ways.

Reply:
IMO, one way or another the data will have to be copied from an `IO` buffer (`InputStream`, `IO` buffer in `ExchangeSource`, etc.) before it can be processed by the engine. The page decoding step seems like a good place to do so. I don't really see why the page decoding should return uncompacted pages if the decoding process requires allocations and memory copying anyway.

Reply:
Agreed, I didn’t notice you had made the same suggestion in another comment.